### PR TITLE
Feat/impersonate users

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

This PR allows ADMINs/Support staff to impersonate a user to assist with any support cases. 

DEMO: https://www.loom.com/share/2b22c355c2c6447f8f8632e2346b499b

TODOs
- [X] Create new provider to login user via impersonation
- [X] Display UI for user when they are impersonating another user
- [X] Allow users to return to them selfs.
- [X] Put code in EE package. 
- [X] Implement a table for audi log of impersonations.

## Type of change
- [X] New feature (non-breaking change which adds functionality)

## How should this be tested?
- [ ] Set a user to be admin within prisma
- [ ] Goto settings>admin and enter a user to impersonate
- [ ] Impersonate user 

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apps/web/components/NavTabs.tsx
Comment (sean-brydon) [inline apps/web/components/NavTabs.tsx:27]: If is admin prop is set render AdminRequired instead of fragment

File: apps/web/components/ui/AdminRequired.tsx
Comment (sean-brydon) [inline apps/web/components/ui/AdminRequired.tsx:11]: Simple way to show content if the user is an admin

File: apps/web/pages/settings/admin.tsx
Comment (sean-brydon) [inline apps/web/pages/settings/admin.tsx:69]: Redirect if you are not an admin

**Review by pumfleet — COMMENTED** [review]
Code looks good overall, will test this in depth due to this being a very security-sensitive PR before approval

File: apps/web/ee/lib/impersonation/ImpersonationProvider.ts
Comment (sean-brydon) [inline apps/web/ee/lib/impersonation/ImpersonationProvider.ts:19]: We should maybe query the user from the session with prisma to ensure it is admin but im sure this is reliable enough as its ran on the server?

**Comment by esprush** [comment]
@sean-brydon @baileypumfleet I understand the value this feature has, but should this include an approval flow from the respective user?
1. admin user requests permission to impersonate
2. on user acceptance, admin user impersonates.

I am curios about this because from user's perspective this would feel like they do not have complete control over the data.

**Comment by sean-brydon** [comment]
> @sean-brydon @baileypumfleet I understand the value this feature has, but should this include an approval flow from the respective user?
> 
> 1. admin user requests permission to impersonate
> 2. on user acceptance, admin user impersonates.
> 
> I am curios about this because from user's perspective this would feel like they do not have complete control over their data.

This is actually something we’ve talked about internally and probably something that needs discussed more with the team - my initial idea was to have the customer select how long they want this feature to be active for 

- user can allow impersonation of their account for the next X amount of time 


@PeerRich @zomars @baileypumfleet

File: apps/web/pages/api/auth/[...nextauth].tsx
Comment (hariombalhara) [inline apps/web/pages/api/auth/[...nextauth].tsx:95]: Would it also work with other auth providers like Google, SAML? I am not sure so just asking this question for confirmation.

**Review by hariombalhara — CHANGES_REQUESTED** [review]
Looks good. Tested locally too.
Except that Google auth and SAML auth providers concern.
Marking "Request Changes" till I have confirmation on that.

File: apps/web/pages/api/auth/[...nextauth].tsx
Comment (sean-brydon) [reply apps/web/pages/api/auth/[...nextauth].tsx:95]: It should - based off they all use the same JWT callback in next-auth. Ill setup them both locally and get back to you

**Comment by sean-brydon** [comment]
@esprush  Thanks for raising your concern this point was raise with our team and we agree - We've added the ability to toggle impersonation on and off into the RFC  as an outstanding TODO #2539

**Comment by PeerRich** [comment]
@sean-brydon what's missing for this to go live? cc @hariombalhara

File: apps/web/ee/lib/impersonation/ImpersonationProvider.ts
Comment (agustif) [reply apps/web/ee/lib/impersonation/ImpersonationProvider.ts:19]: I think this should be safe enough, no need to add an additional database call since session comes from database data anyways? Maybe since it's an internal feature, it wouldn't hurt though, since we actually still use JWT for sessions I guess if someone where to stole the ENV_SECRET for generating those could theoretically generate sessions with ADMIN role on it? @zomars @emrysal @baileypumfleet thoughts?

File: apps/web/ee/lib/impersonation/ImpersonationProvider.ts
Comment (agustif) [inline apps/web/ee/lib/impersonation/ImpersonationProvider.ts:48]: Good thinking on logging the impersonations, maybe in the future we could also add some metadata about actions done during impersionation to improve the -audit- aspects, so if I were to impersonate and change X, y, z, all those actions where to be logged into the impersionating session?

Surely not for this PR/MVP, but just thought for the future

File: apps/web/pages/api/auth/[...nextauth].tsx
Comment (agustif) [reply apps/web/pages/api/auth/[...nextauth].tsx:95]: Yes @hariombalhara I think it should work with those too, since the role level is at the database/model level. As long as we expose the role model to next-auth, all sessions should get this. We need to test it obviously.

File: apps/web/ee/lib/impersonation/ImpersonationProvider.ts
Comment (sean-brydon) [reply apps/web/ee/lib/impersonation/ImpersonationProvider.ts:48]: Im pretty sure this was the plan :)

**Comment by hariombalhara** [comment]
> @sean-brydon what's missing for this to go live? cc @hariombalhara

There was just some confirmation required whether if it would work if user logs in with various auth providers, otherwise it's perfectly good to go.

**Comment by zomars** [comment]
Updating and reviewing now...

File: apps/web/ee/lib/impersonation/ImpersonationProvider.ts
Comment (zomars) [reply apps/web/ee/lib/impersonation/ImpersonationProvider.ts:19]: If someone stole the `ENV_SECRET` we should be in way more bigger trouble than that.

File: apps/web/components/ui/ImpersonatingBanner.tsx
Comment (zomars) [inline apps/web/components/ui/ImpersonatingBanner.tsx]: Let's leave this in one line. Also made it briefer: 
image

**Review by zomars — APPROVED** [review]
Tested and working. Will wait until someone is up to monitor on prod before merging.

**Review by hariombalhara — APPROVED** [review]

**Pull request merged**
